src/functions-and-arrays.js

```diff
@@ -1,41 +1,148 @@
 // Iteration #1: Find the maximum
-function maxOfTwoNumbers() {}
+function maxOfTwoNumbers(num1, num2) {
+  if (num1 > num2) {
+    return num1;
+  } else if (num2 > num1) {
+    return num2;
+  }
+  else{
+    return num1;
+  }
+}
+
+
 
 
 
 // Iteration #2: Find longest word
 const words = ['mystery', 'brother', 'aviator', 'crocodile', 'pearl', 'orchard', 'crackpot'];
 
-function findLongestWord() {}
+function findLongestWord(words) {
+  if (words.length === 0) {
+    return null;
+  } else if (words.length === 1) {
+    return words[0];
+  } else {
+    let longest = words[0];
+    for (let i = 1; i < words.length; i++) {
+      if (words[i].length > longest.length) {
+        longest = words[i];
+      }
+    }
+    return longest;
+  }
+}
+
+
 
 
 
 // Iteration #3: Calculate the sum
 const numbers = [6, 12, 1, 18, 13, 16, 2, 1, 8, 10];
 
-function sumNumbers() {}
+function sumNumbers(numbers) {
+  if (numbers.length === 0) {
+    return 0;
+  }
+  return numbers.reduce((total, num) => total + num, 0);
+}
+
+
 
 
 
 // Iteration #3.1 Bonus:
-function sum() {}
+function sum(arr) {
+  let total = 0;
+  for (let i = 0; i < arr.length; i++) {
+    const element = arr[i];
+    if (typeof element === 'number') {
+      total += element;
+    }
+    else if (typeof element === 'string') {
+      const num = Number(element);
+      if (isNaN(num)) {
+        throw new Error('Unsupported data type in array: string is not a number');
+      }
+      total += num;
+    }
+    else if (typeof element === 'boolean') {
+      if (element) {
+        total += 1;
+      }
+      else if (Array.isArray(element) || typeof element === 'object') {
+        throw new Error ('Unsupported data type in array: object or array');
+      }
+    }
+  }
+    return total;
+}
+
+
 
 
 
 // Iteration #4: Calculate the average
 // Level 1: Array of numbers
 const numbersAvg = [2, 6, 9, 10, 7, 4, 1, 9];
 
-function averageNumbers() {}
+function averageNumbers(numbers) {
+  if (numbers.length === 0) {
+    return null;
+  }
+  let sum = 0;
```

Comment:
We write functions because we can use them later, and since you created a function above to sum numbers, you should reuse it here, instead of writing again these steps to sum the numbers before doing that average.
```
function averageNumbers(numbers) {
  if (!numbers.length) return null;
  let sum = sumNumbers(numbers);
  return sum / numbers.length;
}
```